prov/sockets: Refactored fi_getinfo to handle corner cases

shantonu opened the pull request:
- Added <code>slist_find_first_match()</code> to <code>fi_list.h</code>
- Added <code>sock_matches_interface()</code> that checks if the <code>node</code> or <code>hints->src_addr</code> matches local interfaces
- Minor refactoring of <code>sock_get_list_of_addr</code>
- Return localhost as the last entry in the returned list of interfaces

Fixes #2146 

@jithinjosepkl @shefty @a-ilango

shefty commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +624,55 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
+		ai.ai_family = AF_INET;
+		ai.ai_socktype = SOCK_STREAM;
+
+		ret = getaddrinfo(src_addr, 0, &ai, &rai);
```
This needs to be paired with freeaddrinfo()

shantonu replied:
Missed that! Fix is pushed.

a-ilango commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +623,58 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
+	struct sockaddr_in *addr;
+	int ret;
+
+	if (src_len == sizeof(struct sockaddr_in)) {
```
We might need to handle IPv6 addresses as well.

shefty replied:
I think the socket provider is limited to ipv4 and will require more changes to handle ipv6

jithinjosepkl replied:
yeah, it supports only ipv4 addresses.

a-ilango commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -658,35 +699,44 @@ static int sock_getinfo(uint32_t version, const char *node, const char *service,
+		if (node)
+			ret = sock_matches_interface(addr_list, node, sizeof(node));
+	}
+	else if (hints && hints->src_addr) {
```
nit: "else if" should be in the same line as the closing brace for previous if

shantonu replied:
:+1:

jithinjosepkl commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +623,58 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
+	struct sock_host_list_entry *host_entry;
+	host_entry = container_of(entry, struct sock_host_list_entry, entry);
+
+        if (strcmp(host_entry->hostname, (char *)src_addr) == 0)
```
nit: we could just return directly (strcmp() == 0)

shantonu replied:
:+1:

shantonu commented:
Updated the PR with suggested fixes.

shefty commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +622,57 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
+		ret = getaddrinfo(src_addr, 0, &ai, &rai);
+		if (ret) {
+			SOCK_LOG_DBG("getaddrinfo failed!\n");
+			freeaddrinfo(rai);
```
You need to call freeaddrinfo() if getaddrinfo() succeeds, not fails.

shantonu replied:
:+1:

shefty commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +622,57 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
+	int ret;
+
+	if (src_len == sizeof(struct sockaddr_in)) {
+		addr = (struct sockaddr_in *)src_addr;
```
We shouldn't need a cast here.  You may need to declare addr as const to avoid a warning, and propagate the const through to subcalls.

shantonu replied:
:+1:

shefty commented on include/fi_list.h:
```diff
@@ -187,6 +187,18 @@ static inline struct slist_entry *slist_remove_head(struct slist *list)
 typedef int slist_func_t(struct slist_entry *item, const void *arg);
 
 static inline struct slist_entry *
+slist_find_first_match(struct slist *list, slist_func_t *match, const void *arg)
```
Should probably define list as const

shantonu replied:
:+1:

shefty commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +622,57 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
+	}
+	entry = slist_find_first_match(&addr_list, sock_match_src_addr,
+					inet_ntoa(addr->sin_addr));
+	freeaddrinfo(rai);
```
This should only be called if we call getaddrinfo().  That's only done in the else clause above.

shantonu replied:
Fixed it.

shefty commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -658,35 +697,44 @@ static int sock_getinfo(uint32_t version, const char *node, const char *service,
+	sock_get_list_of_addr(&addr_list);
+	if (flags & FI_SOURCE) {
+		if (node)
+			ret = sock_matches_interface(addr_list, node, sizeof(node));
```
sizeof(node) will be the size of a pointer, not the length of the string.

shantonu replied:
Actually I want pass sizeof(type specifier) which is sizeof(char). Will update.

shantonu commented:
Pushed another round of fixes addressing all the comments.

shantonu commented:
Noticed Mac failures on Travis-CI. Seems to me irrelevant to the PR. Also Travis-CI broadcasted that they have Mac build issues. Not sure if it was sue to system issue.

jithinjosepkl commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -658,35 +696,45 @@ static int sock_getinfo(uint32_t version, const char *node, const char *service,
+
+	if (flags & FI_SOURCE) {
+		if (node)
+			ret = sock_matches_interface(addr_list, node, sizeof(char));
```
It looks a little confusing to use sizeof(char) to determine the address type.

shefty commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +622,57 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
+}
+
+static int sock_matches_interface(struct slist addr_list, const void *src_addr,
+				   size_t src_type_len)
```
As Jithin mentioned below, using a len to indicate the address type is confusing.  Given the structure of the function and that it nearly splits into 2 based on src_type_len, I would just create 2 separate functions and call the correct one from the proper location.

shantonu replied:
I agree, it's a bit confusing. I'll create two functions then.

shantonu commented:
Updates to the PR:
- Fixed a bug in <code>sock_node_matches_interface() </code> which was causing issue in OS X run
- Created two separate methods for matching different address type
- Added logic to detect loopback address and return match without iterating over the local interfaces since it'll always be there.
- Added another commit to fix minor type casting

Travis looks happy as well. Can you please take another look? @jithinjosepkl @shefty

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1380,8 +1380,8 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
 			}
-			memcpy(&_pep->src_addr, result->ai_addr,
-				result->ai_addrlen);
+			memcpy(&_pep->src_addr, (struct sockaddr_in *)result->ai_addr,
```
You don't need a cast for memcpy

shantonu replied:
Oh, ok!

shefty commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +622,62 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
 	return ret;
 }
 
+static int sock_match_src_addr(struct slist_entry *entry, const void *src_addr)
```
Change src_addr from void * to char *.

shefty replied:
Er... you may not be able to - I see that the function is passed as a parameter below

shantonu replied:
Yeah, it's passed in <code>sock_addr_matches_interface()</code> as a function pointer of type <code>slist_func_t(struct slist_entry *item, const void *arg)</code>

jithinjosepkl commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -658,35 +702,45 @@ static int sock_getinfo(uint32_t version, const char *node, const char *service,
+			goto out;
+		}
 	}
+	return 0;
```
Adding return here will result in memory leaks (list entries)

shantonu replied:
It's freed in the while loop in success case and we have a free in error case.

jithinjosepkl commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +622,62 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
+        return (strcmp(host_entry->hostname, (char *) src_addr) == 0);
+}
+
+static int sock_addr_matches_interface(struct slist addr_list, struct sockaddr_in *src_addr)
```
Is there a need for keeping the argument as struct? Cant we use pointer to list here?

shantonu replied:
Sure, will change that.

jithinjosepkl commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -634,6 +622,62 @@ int sock_node_getinfo(const char *node, const char *service, uint64_t flags,
+}
+
+static int sock_node_matches_interface(struct slist addr_list, const char *node)
+{
```
We are passing struct here also.

shantonu replied:
Changed it!

shantonu commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -658,35 +702,45 @@ static int sock_getinfo(uint32_t version, const char *node, const char *service,
+		node = host_entry->hostname;
+		flags |= FI_SOURCE;
+		ret = sock_node_getinfo(node, service, flags, hints, info, &tail);
 		free(host_entry);
```
We are freeing the list entry here @jithinjosepkl

jithinjosepkl replied:
Ok, a cleaner way is to just free the entries before exiting the loop, and then do a break. That can avoid the goto out.

shantonu replied:
Removed the goto and added the free call and return inside the while loop.

shantonu commented:
Addressed all the comments. Is there any other concern?

jithinjosepkl commented:
👍
